src/mpid/ch4/generic/gpu/mpidig_gpu_init.c

```diff
@@ -0,0 +1,118 @@
+      default     : 64
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The size of the buffer in small buffer pool for gpu memory staging.
+
+    - name        : MPIR_CVAR_CH4_GPU_MEDIUM_BUF_POOL_SZ
+      category    : CH4
+      type        : int
+      default     : 32768
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The size of the buffer in medium buffer pool for gpu memory staging.
+
+    - name        : MPIR_CVAR_CH4_GPU_LARGE_BUF_POOL_SZ
+      category    : CH4
+      type        : int
+      default     : 4194304
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The size of the buffer in large buffer pool for gpu memory staging.
+
+    - name        : MPIR_CVAR_CH4_GPU_SMALL_BUF_POOL_NUM
+      category    : CH4
+      type        : int
+      default     : 64
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of the buffers in small buffer pool for gpu memory staging.
+
+    - name        : MPIR_CVAR_CH4_GPU_MEDIUM_BUF_POOL_NUM
+      category    : CH4
+      type        : int
+      default     : 32
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of the buffers in medium buffer pool for gpu memory staging.
+
+    - name        : MPIR_CVAR_CH4_GPU_LARGE_BUF_POOL_NUM
+      category    : CH4
+      type        : int
+      default     : 4
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        The number of the buffers in large buffer pool for gpu memory staging.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+#include "mpidimpl.h"
+#include "ch4r_buf.h"
+#include "mpidig_gpu.h"
+#include "mpidig_gpu_types.h"
+
+static void create_buf_pool(void)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_CREATE_BUF_POOL);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_CREATE_BUF_POOL);
+    int num[MPIDIG_GPU_BUF_POOL_NUM] =
+        { MPIR_CVAR_CH4_GPU_SMALL_BUF_POOL_NUM, MPIR_CVAR_CH4_GPU_MEDIUM_BUF_POOL_NUM,
+        MPIR_CVAR_CH4_GPU_LARGE_BUF_POOL_NUM
+    };
+    size_t size[MPIDIG_GPU_BUF_POOL_NUM] =
+        { MPIR_CVAR_CH4_GPU_SMALL_BUF_POOL_SZ, MPIR_CVAR_CH4_GPU_MEDIUM_BUF_POOL_SZ,
+        MPIR_CVAR_CH4_GPU_LARGE_BUF_POOL_SZ
+    };
+
+    MPIDIU_buf_pool_t **pools = MPIDIG_gpu_global.buf_pool;
+    for (int i = 0; i < MPIDIG_GPU_BUF_POOL_NUM; i++) {
+        pools[i] = MPIDIU_create_buf_pool_internal(num[i], size[i], MPIDIU_BUF_POOL_PINNED, NULL);
+    }
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_GPU_CREATE_BUF_POOL);
```

Comment:
I'm not sure I like this interface.  This hardcodes three pools (small, medium, large) and gives the caller no flexibility for the size.  There's just one global setting (through CVARs) for the size.  What if intranode and internode communication need two different sizes?  It'd be cleaner if the "buffer pool" utility be treated as a separate orthogonal class that simply creates a "pool object" with whatever parameters the user provides (e.g., size, number of buffers, etc.).

Here's what yaksa does, for example: https://github.com/pmodels/yaksa/blob/master/src/util/yaksu_pool.c